Keep subclassing in apply

jaumebonet opened the pull request:
When generating new objects on apply, it calls `self.obj._constructor` and `self.obj._constructior_sliced` instead of:  `DataFrame` and `Series`; keeping subclassing.

- [x] closes #19822
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] keeps subclass inside `DataFrame.apply`

jaumebonet committed:
Keep subclassing in apply
When generating new objects from apply, it calls the
self.obj._constructor, self.obj._constructior_sliced instead of
DataFrame, Series; keeping subclassing.

pep8speaks commented:
Hello @jaumebonet! Thanks for submitting the PR.

 - In the file [`pandas/core/apply.py`](https://github.com/pandas-dev/pandas/blob/0df868934d8d205c27e5014dd5600d6595d11cf6/pandas/core/apply.py), following are the PEP8 issues :

> [Line 178:80](https://github.com/pandas-dev/pandas/blob/0df868934d8d205c27e5014dd5600d6595d11cf6/pandas/core/apply.py#L178): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 225:80](https://github.com/pandas-dev/pandas/blob/0df868934d8d205c27e5014dd5600d6595d11cf6/pandas/core/apply.py#L225): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)

jaumebonet committed:
apply changes due to tests break

jaumebonet commented:
Hi @chris-b1,
Thanks!
Yes, I'm trying to pinpoint what exactly has changed... Clearly it has to do with the fact that now it returns `SparseSeries` instead of `Series`, but I'm not familiar with `SparseSeries`, so I'm not sure what difference does that make.

jaumebonet committed:
FrameColumnApply.fix errors in wrap_result_for_axis

jreback commented:
can you add some tests in pandas/tests/frame/test_sub_class (and for Series as well)

jreback commented:
also pls add a release note, other enhancements section is ok

jaumebonet committed:
Adding test and release note

jaumebonet commented:
Hi @jreback,
I've added a test to check if subclass of `Series` are kept when using apply on `DataFrame`.
`Series` have they own apply function, so the changes do not affect how `Series.apply` behave. 

As per your suggestion, I also added a bullet point in the _other enhancement_ section.

jaumebonet committed:
fix typo

jreback commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -295,6 +295,7 @@ Other Enhancements
 - ``IntervalIndex.astype`` now supports conversions between subtypes when passed an ``IntervalDtype`` (:issue:`19197`)
 - :class:`IntervalIndex` and its associated constructor methods (``from_arrays``, ``from_breaks``, ``from_tuples``) have gained a ``dtype`` parameter (:issue:`19262`)
 - Added :func:`SeriesGroupBy.is_monotonic_increasing` and :func:`SeriesGroupBy.is_monotonic_decreasing` (:issue:`17015`)
+- :func:``DataFrame.apply`` keeps provides the specified ``Series`` subclass when `DataFrame._constructor_sliced`` is defined (:issue:`19822`)
```
don't use a private property here, just say when ``Series`` and ``DataFrame``` are subclassed

jreback commented on pandas/tests/frame/test_subclass.py:
```diff
@@ -514,3 +514,19 @@ def test_subclassed_wide_to_long(self):
+            ['Mary', 'Bo', 'weight', 150]],
+            columns=['first', 'last', 'variable', 'value'])
+
+        df.apply(lambda x: check_row_subclass(x))
```
you need to check the result of this using ``tm.assert_frame_equal``, just use a simple example of an applied function, testing both returning a list, e.g. ``lambda x: [1, 2, 3]`` and a sub-classed series

jaumebonet committed:
update test_subclassed_apply and release note

jaumebonet committed:
pep8 styling

jaumebonet committed:
Merge branch 'master' into apply-subclass

TomAugspurger commented on pandas/tests/frame/test_subclass.py:
```diff
@@ -514,3 +514,55 @@ def test_subclassed_wide_to_long(self):
+            [1, 2, 3],
+            [1, 2, 3]])
+
+        expected3 = DesignSeries([[1, 2, 3], [1, 2, 3], [1, 2, 3], [1, 2, 3]])
```
Looks like this needs to be updated.

jaumebonet committed:
test fix

jaumebonet committed:
fix test

jaumebonet committed:
fixed lint "line too long" error in travis ci

jreback commented on pandas/tests/frame/test_subclass.py:
```diff
@@ -514,3 +514,59 @@ def test_subclassed_wide_to_long(self):
+        expected3 = tm.SubclassedSeries([
+            [1, 2, 3],
+            [1, 2, 3],
+            [1, 2, 3],
```
rather than numbering these, can you just do one after the other and use 
```
result = 
expected = 
```

jaumebonet committed:
fix test styling

jaumebonet committed:
fixed test typo

jorisvandenbossche commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -295,8 +295,10 @@ Other Enhancements
 - ``IntervalIndex.astype`` now supports conversions between subtypes when passed an ``IntervalDtype`` (:issue:`19197`)
 - :class:`IntervalIndex` and its associated constructor methods (``from_arrays``, ``from_breaks``, ``from_tuples``) have gained a ``dtype`` parameter (:issue:`19262`)
 - Added :func:`SeriesGroupBy.is_monotonic_increasing` and :func:`SeriesGroupBy.is_monotonic_decreasing` (:issue:`17015`)
+- :func:``DataFrame.apply`` keeps the specified ``Series`` subclass when ``Series`` and ``DataFrame`` subclasses are defined (:issue:`19822`)
```
I find this not so clear. It is about the Series type that is passed to the function?

jaumebonet replied:
Yes.  
When `apply`ing a function to a subclassed `DataFrame` I realised I could not call the specific methods of the subclassed `Series` that this `DataFrame` was supossed to generate  through `_constructor_sliced`. As far as I understood, the subclassing of the `Series` was lost during `apply`.  
With these changes the subclass is now kept inside `apply`, so method of the expected `Series` subclass can be used.

jorisvandenbossche replied:
OK, can you try to clarify the text a bit? Eg something like "For subclassed DataFrames, DataFrame.apply will now preserve the Series subclass (if defined) when passing the data to the applied function" (but adapt as you like)

jaumebonet replied:
sure! I like it as this! :)
will change it!

jaumebonet committed:
fix docs

jreback commented:
thanks!
